File: tests/iperf3/kind/simple_test.py
################################################################
#Name: Simple iperf3  test
#Desc: create 2 kind clusters :
# 1) MBG and iperf3 client
# 2) MBG and iperf3 server    
###############################################################
#!/usr/bin/env python3
import os,time
import subprocess as sp
import sys
import argparse


proj_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname( os.path.abspath(__file__)))))
sys.path.insert(0,f'{proj_dir}')

from tests.utils.mbgAux import runcmd, runcmdb, printHeader, waitPod, getPodName, getMbgPorts,buildMbg,buildMbgctl,getPodIp,getPodNameIp
from tests.iperf3.kind.connect_mbgs import connectMbgs, sendHello
from tests.iperf3.kind.iperf3_service_create import setIperf3client, setIperf3Server
from tests.iperf3.kind.iperf3_service_expose import exposeService
from tests.iperf3.kind.iperf3_service_get import getService
from tests.iperf3.kind.iperf3_client_start import directTestIperf3,testIperf3Client
from tests.iperf3.kind.apply_policy import applyPolicy

from tests.utils.kind.kindAux import useKindCluster, getKindIp,startKindClusterMbg

############################### MAIN ##########################
if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Description of your program')
    parser.add_argument('-d','--dataplane', help='choose which dataplane to use mtls/tcp', required=False, default="mtls")
    parser.add_argument('-c','--cni', help='Which cni to use default(kindnet)/flannel/calico/diff (different cni for each cluster)', required=False, default="default")

    args = vars(parser.parse_args())

    printHeader("\n\nStart Kind Test\n\n")
    printHeader("Start pre-setting")

    dataplane = args["dataplane"]
    cni = args["cni"]
    #MBG1 parameters 
    mbg1DataPort    = "30001"
    mbg1cPort       = "30443"
    mbg1cPortLocal  = "8443"
    mbg1crtFlags    = f"--rootCa ./mtls/ca.crt --certificate ./mtls/mbg1.crt --key ./mtls/mbg1.key"  if dataplane =="mtls" else ""
    mbg1Name        = "mbg1"
    mbgctl1Name     = "mbgctl1"
    mbg1cni         = cni 
    srcSvc          = "iperf3-client"
    srck8sSvcPort   = "5000"
    
    #MBG2 parameters 
    mbg2DataPort    = "30001"
    mbg2cPort       = "30443"
    mbg2cPortLocal  = "8443"
    mbg2crtFlags    = f"--rootCa ./mtls/ca.crt --certificate ./mtls/mbg2.crt --key ./mtls/mbg2.key"  if dataplane =="mtls" else ""
    mbg2Name        = "mbg2"
    mbgctl2Name     = "mbgctl2"
    mbg2cni         = "flannel" if cni == "diff" else cni
    destSvc         = "iperf3-server"
    iperf3DestPort  = "30001"
    
        
    #folders
    folCl=f"{proj_dir}/tests/iperf3/manifests/iperf3-client"
    folSv=f"{proj_dir}/tests/iperf3/manifests/iperf3-server"
    
    print(f'Working directory {proj_dir}')
    os.chdir(proj_dir)
    
    ### clean 
    print(f"Clean old kinds")
    os.system("make clean-kind-iperf3")
    
    ### build docker environment 
    printHeader(f"Build docker image")
    os.system("make docker-build")
    
    
    ### Build MBG in Kind clusters environment 
    startKindClusterMbg(mbg1Name, mbgctl1Name, mbg1cPortLocal, mbg1cPort, mbg1DataPort, dataplane ,mbg1crtFlags, cni=mbg1cni)        
    startKindClusterMbg(mbg2Name, mbgctl2Name, mbg2cPortLocal, mbg2cPort, mbg2DataPort, dataplane ,mbg2crtFlags, cni=mbg2cni)        
      
    ###get mbg parameters
    useKindCluster(mbg1Name)
    mbg1Pod, _           = getPodNameIp("mbg")
    mbg1Ip               = getKindIp("mbg1")
    mbgctl1Pod, mbgctl1Ip= getPodNameIp("mbgctl")
    useKindCluster(mbg2Name)
    mbg2Pod, mbg2Ip       = getPodNameIp("mbg")
    mbgctl2Pod, mbgctl2Ip = getPodNameIp("mbgctl")
    destkindIp=getKindIp(mbg2Name)

    
    # Add MBG Peer
    useKindCluster(mbg2Name)
    printHeader("Add MBG2 peer to MBG1")
    connectMbgs(mbg2Name, mbgctl2Name, mbgctl2Pod, mbg1Name, mbg1Ip, mbg1cPort)

    # Send Hello
    sendHello(mbgctl2Pod, mbgctl2Name)        

    
    # Set service iperf3-client in MBG1
    setIperf3client(mbg1Name, mbgctl1Name, srcSvc)
    
    # Set service iperf3-server in MBG2
    setIperf3Server(mbg2Name, mbgctl2Name,destSvc)

    #Expose destination service
    exposeService(mbg2Name, mbgctl2Name, destSvc)

    #Get services
    getService(mbg1Name, mbgctl1Name)
    
    #Testing
    printHeader("\n\nStart Iperf3 testing")
    useKindCluster(mbg2Name)
    waitPod("iperf3-server")
    
    #Test MBG1
    directTestIperf3(mbg1Name, srcSvc, destSvc,destkindIp,iperf3DestPort)
    testIperf3Client(mbg1Name,srcSvc,destSvc)

File: tests/iperf3/kind/simple-test.go
// ###############################################################
// Name: Simple iperf3  test
// Desc: create 2 kind clusters :
// 1) MBG and iperf3 client
// 2) MBG and iperf3 server
// ##############################################################
package main

import (
	"os"
	"strconv"
	"time"

	log "github.com/sirupsen/logrus"
	"github.ibm.com/mbg-agent/cmd/gwctl/config"
	"github.ibm.com/mbg-agent/pkg/admin"
	api "github.ibm.com/mbg-agent/pkg/api/admin"
	mbgAux "github.ibm.com/mbg-agent/tests/utils"
	kindAux "github.ibm.com/mbg-agent/tests/utils/kind"
)

const (
	mbgCaCrt = "./mtls/ca.crt"
	//MBG1 parameters
	mbg1DataPort   = "30001"
	mbg1cPort      = 30443
	mbg1cPortLocal = "443"
	mbg1crt        = "./mtls/mbg1.crt"
	mbg1key        = "./mtls/mbg1.key"
	mbg1Name       = "mbg1"
	gwctl1Name     = "gwctl1"
	mbg1cni        = "default"
	srcSvc         = "iperf3-client"

	//MBG2 parameters
	mbg2DataPort   = "30001"
	mbg2cPort      = 30443
	mbg2cPortLocal = "443"
	mbg2crt        = "./mtls/mbg2.crt"
	mbg2key        = "./mtls/mbg2.key"
	mbg2Name       = "mbg2"
	gwctl2Name     = "gwctl2"
	mbg2cni        = "default"
	destSvc        = "iperf3-server"
	destPort       = uint16(5000)
	kindDestPort   = "30001"
)

var (
	mtlsFolder string = mbgAux.ProjDir + "/tests/utils/"
	folCl      string = mbgAux.ProjDir + "/tests/iperf3/manifests/iperf3-client"
	folSv      string = mbgAux.ProjDir + "/tests/iperf3/manifests/iperf3-server"
)

func main() {
	// call a Python function
	dataplane := "mtls"
	nologfile := false
	mbgAux.SetLog()
	log.Println("Working directory", mbgAux.ProjDir)
	//exec.chdir(proj_dir)
	//clean
	log.Print("Clean old kinds")
	mbgAux.RunCmd("make clean-kind")

	// build docker environment
	mbgAux.PrintHeader("Build docker image")
	mbgAux.RunCmd("make docker-build")
	kindAux.CreateKindMbg(mbg1Name, dataplane, nologfile)
	kindAux.CreateKindMbg(mbg2Name, dataplane, nologfile)

	// //get parameters
	mbg1Ip, _ := kindAux.GetKindIp(mbg1Name)
	mbg2Ip, _ := kindAux.GetKindIp(mbg2Name)

	//set gwctl
	gwctl1, err := admin.NewClient(config.ClientConfig{
		ID:        gwctl1Name,
		GwIP:      mbg1Ip + ":" + strconv.Itoa(mbg1cPort),
		CaFile:    mtlsFolder + mbgCaCrt,
		CertFile:  mtlsFolder + mbg1crt,
		KeyFile:   mtlsFolder + mbg1key,
		Dataplane: dataplane,
	})
	if err != nil {
		log.Error(err)
		os.Exit(1)
	}
	gwctl2, err := admin.NewClient(config.ClientConfig{
		ID:        gwctl2Name,
		GwIP:      mbg2Ip + ":" + strconv.Itoa(mbg2cPort),
		CaFile:    mtlsFolder + mbgCaCrt,
		CertFile:  mtlsFolder + mbg2crt,
		KeyFile:   mtlsFolder + mbg2key,
		Dataplane: dataplane,
	})

	if err != nil {
		log.Error(err)
		os.Exit(1)
	}
	//Add Peer
	mbgAux.PrintHeader("Add peers and send hello")
	gwctl1.CreatePeer(api.Peer{Name: mbg2Name, Spec: api.PeerSpec{Gateways: []api.Endpoint{{Host: mbg2Ip, Port: mbg2cPort}}}})

	//Set iperf3 client
	mbgAux.PrintHeader("Add iperf3 client")
	kindAux.CreateServiceInKind(mbg1Name, srcSvc, "mlabbe/iperf3", folCl+"/"+srcSvc+".yaml")
	srcSvcPod, _ := mbgAux.GetPodNameIp(srcSvc)
	//gwctl1.AddService(srcSvc, "", "", "iperf3 client") //Allow to use all by default

	//Set iperf3 server
	mbgAux.PrintHeader("Add iperf3 server")
	kindAux.CreateServiceInKind(mbg2Name, destSvc, "mlabbe/iperf3", folSv+"/iperf3.yaml")
	destSvcPod, destSvcIP := mbgAux.GetPodNameIp(destSvc)

	gwctl2.CreateExportService(api.Export{Name: destSvc, Spec: api.ExportSpec{Service: api.Endpoint{Host: destSvcIP, Port: destPort}}})
	log.Println(srcSvcPod, destSvcPod)

	//Expose service
	mbgAux.PrintHeader("Start expose")
	kindAux.UseKindCluster(mbg2Name)
	gwctl2.CreateImportService(api.Import{Name: destSvc, Spec: api.ImportSpec{Service: api.Endpoint{Host: destSvc, Port: destPort}}})

	//bindK8sSvc()
	mbgAux.PrintHeader("Bind a service")
	kindAux.UseKindCluster(mbg1Name)
	gwctl1.CreateBinding(api.Binding{Spec: api.BindingSpec{Import: destSvc, Peer: mbg2Name}})
	time.Sleep(5 * time.Second)

	//iperf3test
	mbgAux.RunCmdNoPipe("kubectl exec -i " + srcSvcPod + " -- iperf3 -c " + destSvc + " -p " + strconv.Itoa(int(destPort)))

}

// ############################### MAIN ##########################
// if __name__ == "__main__":
//     parser = argparse.ArgumentParser(description='Description of your program')
//     parser.add_argument('-d','--dataplane', help='choose which dataplane to use mtls/tcp', required=False, default="mtls")
//     parser.add_argument('-c','--cni', help='Which cni to use default(kindnet)/flannel/calico/diff (different cni for each cluster)', required=False, default="default")
